File: 3470.permutations-iv.java
#
# @lc app=leetcode id=3470 lang=java
#
# [3470] Permutations IV
#

# @lc code=start
class Solution {
    public int[] permute(int n, long k) {
        if (n == 0) return new int[0];
        int numOdds = (n + 1) / 2;
        int numEvens = n / 2;
        long c1 = countWays(numOdds, numEvens, true, k);
        long c2 = countWays(numOdds, numEvens, false, k);
        if (c1 < k && c2 < k && c1 + c2 < k) {
            return new int[0];
        }
        boolean[] used = new boolean[n + 1];
        int[] result = new int[n];
        int remOdds = numOdds;
        int remEvens = numEvens;
        long remK = k;
        boolean lastOdd = false;
        boolean hasPrev = false;
        for (int pos = 0; pos < n; ++pos) {
            boolean needOdd = !lastOdd;
            boolean chosen = false;
            for (int cand = 1; cand <= n; ++cand) {
                if (used[cand]) continue;
                boolean isOdd = (cand % 2 == 1);
                if (hasPrev && isOdd != needOdd) continue;
                int nrO = remOdds - (isOdd ? 1 : 0);
                int nrE = remEvens - (isOdd ? 0 : 1);
                if (nrO < 0 || nrE < 0) continue;
                boolean nextNeedOdd = !isOdd;
                long ways = countWays(nrO, nrE, nextNeedOdd, remK);
                if (ways >= remK) {
                    result[pos] = cand;
                    used[cand] = true;
                    lastOdd = isOdd;
                    hasPrev = true;
                    if (isOdd) --remOdds;
                    else --remEvens;
                    chosen = true;
                    break;
                } else {
                    remK -= ways;
                }
            }
            if (!chosen) {
                return new int[0];
            }
        }
        return result;
    }

    private static long countWays(int odds, int evens, boolean startOdd, long cap) {
        long K = cap;
        int total = odds + evens;
        if (total == 0) return 1L;
        int oddSlots = startOdd ? (total + 1) / 2 : total / 2;
        int evenSlots = total - oddSlots;
        if (oddSlots != odds || evenSlots != evens) return 0L;
        long factO = 1L;
        for (int i = 2; i <= odds; ++i) {
            if (factO > K / i) return K + 1;
            factO *= i;
        }
        long factE = 1L;
        for (int i = 2; i <= evens; ++i) {
            if (factE > K / i) return K + 1;
            factE *= i;
        }
        if (factE > 0 && factO > K / factE) {
            return K + 1;
        }
        return factO * factE;
    }
}
# @lc code=end

File: 3470.permutations-iv.go
#
# @lc app=leetcode id=3470 lang=golang
#
# [3470] Permutations IV
#

# @lc code=start
func permute(n int, k int64) []int {
	used := make([]bool, n+1)
	remOddCnt := (n + 1) / 2
	remEvenCnt := n / 2

	fact := [21]uint64{}
	fact[0] = 1
	for i := 1; i <= 20; i++ {
		fact[i] = fact[i-1] * uint64(i)
	}

	currK := uint64(k)
	var needOdd bool
	res := make([]int, n)

	for pos := 0; pos < n; pos++ {
		found := false
		m := n - pos
		for cand := 1; cand <= n; cand++ {
			if used[cand] {
				continue
			}
			isOddCand := (cand % 2 == 1)
			if pos > 0 && isOddCand != needOdd {
				continue
			}
			do := 0
			if isOddCand {
				do = 1
			}
			de := 1 - do
			if remOddCnt < do || remEvenCnt < de {
				continue
			}
			ra := remOddCnt - do
			rb := remEvenCnt - de
			s := m - 1
			nextOdd := !isOddCand
			nos := numOddSlots(s, nextOdd)
			nes := s - nos
			if ra != nos || rb != nes {
				continue
			}

			large := ra > 20 || rb > 20
			if large {
				res[pos] = cand
				used[cand] = true
				if isOddCand {
					remOddCnt--
				} else {
					remEvenCnt--
				}
				needOdd = !isOddCand
				found = true
				break
			} else {
				fa := fact[ra]
				fb := fact[rb]
				if currKLeCnt(currK, fa, fb) {
					res[pos] = cand
					used[cand] = true
					if isOddCand {
						remOddCnt--
					} else {
						remEvenCnt--
					}
					needOdd = !isOddCand
					found = true
					break
				} else {
					fcnt := float64(fa) * float64(fb)
					ucnt := uint64(fcnt)
					currK -= ucnt
				}
			}
		}
		if !found {
			return []int{}
		}
	}

	return res
}

func numOddSlots(length int, startOdd bool) int {
	if startOdd {
		return (length + 1) / 2
	}
	return length / 2
}

func currKLeCnt(ck uint64, fa uint64, fb uint64) bool {
	if fb == 0 {
		return ck <= 1
	}
	ceilDiv := (ck + fb - 1) / fb
	return fa >= ceilDiv
}
# @lc code=end